server/sonar-process-monitor/src/main/java/org/sonar/application/process/ProcessLauncherImpl.java

```diff
@@ -0,0 +1,256 @@
+    try {
+      ProcessCommands commands = allProcessesCommands.createAfterClean(javaCommand.getProcessId().getIpcIndex());
+
+      ProcessBuilder processBuilder = create(javaCommand);
+      LOG.info("Launch process[{}]: {}", javaCommand.getProcessId().getKey(), String.join(" ", processBuilder.command()));
+      process = processBuilder.start();
+      return new ProcessCommandsProcessMonitor(process, commands);
+    } catch (Exception e) {
+      // just in case
+      if (process != null) {
+        process.destroyForcibly();
+      }
+      throw new IllegalStateException(format("Fail to launch process [%s]", javaCommand.getProcessId().getKey()), e);
+    }
+  }
+
+  private ProcessBuilder create(EsCommand esCommand) {
+    List<String> commands = new ArrayList<>();
+    commands.add(esCommand.getExecutable().getAbsolutePath());
+    commands.addAll(esCommand.getEsOptions());
+
+    return create(esCommand, commands);
+  }
+
+  private ProcessBuilder create(JavaCommand javaCommand) {
+    List<String> commands = new ArrayList<>();
+    commands.add(buildJavaPath());
+    commands.addAll(javaCommand.getJavaOptions());
+    // TODO warning - does it work if temp dir contains a whitespace ?
+    // TODO move to CommandFactory ?
+    commands.add(format("-Djava.io.tmpdir=%s", tempDir.getAbsolutePath()));
+    commands.addAll(buildClasspath(javaCommand));
+    commands.add(javaCommand.getClassName());
+    commands.add(buildPropertiesFile(javaCommand).getAbsolutePath());
+
+    return create(javaCommand, commands);
+  }
+
+  private ProcessBuilder create(AbstractCommand<?> javaCommand, List<String> commands) {
+    ProcessBuilder processBuilder = processBuilderSupplier.get();
+    processBuilder.command(commands);
+    processBuilder.directory(javaCommand.getWorkDir());
+    processBuilder.environment().putAll(javaCommand.getEnvVariables());
+    processBuilder.redirectErrorStream(true);
+    return processBuilder;
+  }
+
+  private static String buildJavaPath() {
+    String separator = System.getProperty("file.separator");
+    return new File(new File(System.getProperty("java.home")), "bin" + separator + "java").getAbsolutePath();
+  }
+
+  private static List<String> buildClasspath(JavaCommand javaCommand) {
+    String pathSeparator = System.getProperty("path.separator");
+    return Arrays.asList("-cp", String.join(pathSeparator, javaCommand.getClasspath()));
+  }
+
+  private File buildPropertiesFile(JavaCommand javaCommand) {
+    File propertiesFile = null;
+    try {
+      propertiesFile = File.createTempFile("sq-process", "properties", tempDir);
+      Properties props = new Properties();
+      props.putAll(javaCommand.getArguments());
+      props.setProperty(PROPERTY_PROCESS_KEY, javaCommand.getProcessId().getKey());
+      props.setProperty(PROPERTY_PROCESS_INDEX, Integer.toString(javaCommand.getProcessId().getIpcIndex()));
+      // FIXME is it the responsibility of child process to have this timeout (too) ?
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)